core/object/worker_thread_pool.cpp

```diff
@@ -150,88 +132,163 @@ void WorkerThreadPool::_process_task(Task *p_task) {
 
 		task_mutex.lock();
 		p_task->completed = true;
-		for (uint8_t i = 0; i < p_task->waiting; i++) {
-			p_task->done_semaphore.post();
+		p_task->pool_thread_index = -1;
+		if (p_task->waiting_user) {
+			p_task->done_semaphore.post(p_task->waiting_user);
 		}
-		if (!use_native_low_priority_threads) {
-			p_task->pool_thread_index = -1;
+		// Let awaiters know.
+		for (uint32_t i = 0; i < threads.size(); i++) {
+			if (threads[i].awaited_task == p_task) {
+				threads[i].cond_var.notify_one();
+			}
 		}
-		task_mutex.unlock(); // Keep mutex down to here since on unlock the task may be freed.
 	}
 
-	// Task may have been freed by now (all callers notified).
-	p_task = nullptr;
+	p_task = nullptr; // May have been freed by now.
 
-	if (!use_native_low_priority_threads) {
-		bool post = false;
-		task_mutex.lock();
+	{
 		ThreadData &curr_thread = threads[pool_thread_index];
-		curr_thread.current_low_prio_task = prev_low_prio_task;
+		curr_thread.current_task = prev_task;
 		if (low_priority) {
 			low_priority_threads_used--;
-			low_priority_tasks_running--;
-			// A low prioriry task was freed, so see if we can move a pending one to the high priority queue.
+		} else {
 			if (_try_promote_low_priority_task()) {
-				post = true;
-			}
-
-			if (low_priority_tasks_awaiting_others == low_priority_tasks_running) {
-				_prevent_low_prio_saturation_deadlock();
+				if (prev_task) { // Otherwise, this thread will catch it.
+					_notify_threads(&curr_thread, 1, 0);
+				}
 			}
 		}
 		task_mutex.unlock();
-		if (post) {
-			task_available_semaphore.post();
-		}
 	}
+
+	set_current_thread_safe_for_nodes(safe_for_nodes_backup);
 }
 
 void WorkerThreadPool::_thread_function(void *p_user) {
+	ThreadData &thread_data = singleton->threads[singleton->get_thread_index()];
```

Comment:
This fails if the return is `-1`, when the thread id is not present

Reply:
This case has turned out to be very interesting! At first, I thought that, since this function only serves as the entry point of the thread, there's no possibility the id is not found. Then, I realized that, given how the ids map and the threads array are initialized, there is a race condition. Finally, I decided to directly pass the pointer to the relevant thread data structure via user data. Then, and this is the amusing thing, I found that the threads are already created with `&threads[i]` passed as user data, and that was the case when the WTP was first written, even if it wasn't used then! It's as if Juan had left it for me in the future.

Reply:
The adventures in code! Good detective work!